### PR TITLE
feat(reporters): show full test suite when testing 1 spec file at a time

### PR DESCRIPTION
**Pull request by Dunqing**
close: #3266

### CODE REVIEW

File: test/reporters/tests/default.test.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { resolve } from 'pathe'
+import { describe, expect, test } from 'vitest'
+import { runVitest } from '../../test-utils'
+
+const root = resolve(__dirname, '../fixtures')
+const run = async (fileFilter: string[]) => await runVitest({ root, reporters: 'default', config: false }, fileFilter)
+
+describe.skipIf(process.env.CI)('default reporter', async () => {
```
Comment (sheremet-va) [inline test/reporters/tests/default.test.ts:8]: Do I even need to explain why you shouldn't skip tests in CI?

Comment (Dunqing) [reply test/reporters/tests/default.test.ts:8]: This function only works with TTY, but CI will cause isTTY to be false

Comment (Dunqing) [reply test/reporters/tests/default.test.ts:8]: Aha, yes, I forgot that I could use this very useful method.😅

**Review by sheremet-va — APPROVED** [review]

File: packages/vitest/src/node/reporters/default.ts [hunk 2]
```diff
@@ -8,6 +8,11 @@ export class DefaultReporter extends BaseReporter {
   renderer?: ReturnType<typeof createListRenderer>
   rendererOptions: ListRendererOptions = {} as any
 
+  onPathsCollected(paths: string[] = []) {
+    if (paths.length === 1)
+      this.rendererOptions.renderSucceed = true
+  }
```
Comment (sheremet-va) [inline packages/vitest/src/node/reporters/default.ts:14]: We should probably undo the change if watch mode triggers rerun again but with >1 tests

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vitest/src/node/reporters/default.ts (hunk 2 above)
Comment (Dunqing) [reply packages/vitest/src/node/reporters/default.ts:22]: Done, Thank you for your review!

**Comment by sheremet-va** [comment]
Looks like tests are failing now

**Comment by Dunqing** [comment]
The tests looks good now.

File: test/reporters/tests/default.test.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    await vitest.waitForStdout('Input filename pattern')
+    vitest.write('b\n')
+    await vitest.waitForStdout('Waiting for file changes')
+    expect(vitest.stdout).toContain('RERUN')
+    expect(vitest.stdout).toContain('b1.test.ts')
+    expect(vitest.stdout).toContain('b2.test.ts')
+    expect(vitest.stdout).not.toContain('nested b1 test')
+    expect(vitest.stdout).not.toContain('b1 test')
+    expect(vitest.stdout).not.toContain('b2 test')
+  }, {
```
Comment (sheremet-va) [inline test/reporters/tests/default.test.ts:55]: This only proves the test is unstable. Maybe it's better to add some kind of waiting if it's the problem? It usually fails on line 53 - why?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: test/reporters/tests/default.test.ts (hunk 3 above)
Comment (Dunqing) [reply test/reporters/tests/default.test.ts]: Yes, it does seem to be unstable in CI, but I'm not sure what's causing it yet and I need to spend some time checking it out.

**Review by sheremet-va — APPROVED** [review]

**Comment by Dunqing** [comment]
> Looks like the test is still very unstable. Please, have a look when you have more time @Dunqing

Ok, I will do it later.

**Comment by sheremet-va** [comment]
> > Looks like the test is still very unstable. Please, have a look when you have more time @Dunqing
> 
> Ok, I will do it later.

Another issue I see now is that `console.log` is hidden at the top (before the screen) if there are a lot of tests inside a file which leads to degraded user experience in my opinion.

**Comment by Dunqing** [comment]
> Another issue I see now is that `console.log` is hidden at the top (before the screen) if there are a lot of tests inside a file which leads to degraded user experience in my opinion.

This behaviour should be the same as `jest`, If I cared about the contents of the `console.log`, I think I would use `test.only`. 

Or do we need to make sure that the number of tests in the file < xxxx before enabling this behaviour?

**Comment by sheremet-va** [comment]
> This behaviour should be the same as `jest`,

Is it the same as jest? This test file runs quite differently:

```ts
it.each(new Array(100).fill(0).map((_, i) => i))('test', async (num) => {
  expect(true).toBe(true)
  await new Promise(resolve => setTimeout(resolve, num))
})
```

Vitest shows all 100 tests, but jest only shows the file name

**Comment by Dunqing** [comment]
> > This behaviour should be the same as `jest`,
> 
> Is it the same as jest? This test file runs quite differently:
> 
> ```ts
> it.each(new Array(100).fill(0).map((_, i) => i))('test', async (num) => {
>   expect(true).toBe(true)
>   await new Promise(resolve => setTimeout(resolve, num))
> })
> ```
> 
> Vitest shows all 100 tests, but jest only shows the file name

image

I tested the code you provided with jest and at the end of the run, all the tests are printed.

**Comment by Dunqing** [comment]
> > I tested the code you provided with jest and at the end of the run, all the tests are printed.
> 
> Yes, in the end, but not _during_

Oh, yes, It looks like this is the wrong implementation and needs to be changed to print only at the end of the test.

**Comment by Dunqing** [comment]
@sheremet-va 

When I revert this pr and tested again using `vitest` with the code you provided, the results were the same, `vitest` will show all 100 tests. I guess this is because `it.each` is handled differently in `jest` and `vitest`?

**Comment by sheremet-va** [comment]
> @sheremet-va
> 
> When I revert this pr and tested again using `vitest` with the code you provided, the results were the same, `vitest` will show all 100 tests. I guess this is because `it.each` is handled differently in `jest` and `vitest`?

I don't know how it's handled in jest. In vitest `test.eash` is the same as calling `test` function 100 times. I guess if behavior didn't change then it should be ok for the time being?..

**Comment by sheremet-va** [comment]
Tets fails too often tho, we might have to revert this PR before release. Do you have plans to fix the issue with a flaky test?

**Comment by sheremet-va** [comment]
> I'm working on this, I'm still not sure what's causing the different results, can't we really add retries for it?

Retry just halves the time it will fail, not fix the issue

**Comment by ghiscoding** [comment]
@Dunqing just curious was this PR issue fixed or is it still falling because that caused some unit tests to fail right? I can't wait to see this in action though 😉

**Comment by Dunqing** [comment]
> @Dunqing just curious was this PR issue fixed or is it still falling because that caused some unit tests to fail right? I can't wait to see this in action though 😉

It has been fixed, and we should see it in the next release.

**Comment by ghiscoding** [comment]
@Dunqing  sweet thanks, however just to let you know that the issue #3266 was actually reopened by @sheremet-va, someone might need to recheck if it can be closed again :)
